torchao/prototype/quantized_training/int8_mixed_precision.py

```diff
@@ -0,0 +1,203 @@
+from typing import Any, NamedTuple, Optional, Tuple
+
+import torch
+from torch import Tensor, nn
+from torch.utils._python_dispatch import return_and_correct_aliasing
+from torch.utils._triton import has_triton
+
+from torchao.dtypes.utils import _dispatch__torch_dispatch__, _dispatch__torch_function__, _implements
+
+from .int8 import quantize_int8_rowwise
+
+if has_triton():
+    from .int8_mm import int8_mm_dequant
```

Comment:
have you tried using `torch._int_mm` with torch.compile in `max-autotune` mode to generate the matmul?

Reply:
I haven't. I can try.

A question. Can `torch_int_mm()` generate INT8 matmul fused with dequant? If yes, what is the proper way to write matmul that torch.compile will codegen it correctly? I see some versions of this in dynamo template but not very sure under which condition it will be codegen-ed.

Some problems I often face:
- `max-autotune` can take very long, especially for training.
- triton codegen can be finicky i.e. it may or may not do what we expect. Feels like sometimes it's clearer to manually write out the triton kernel (though we may miss out some other fusion opportunities)

Reply:
makes sense.  To scale to all preceding/subsequent ops, IMO compiler is the only scalable option.  If you are ok with just covering the matmul and not fusing preceding/subsequent ops, manual kernel works great.  I was more just wondering if you have have tried using the compiler.

Reply:
I was playing around a bit with `torch._int_mm()` with max-autotune to codegen the fused matmul. Some observations:
- Seems like currently inductor can only fuse 1 epilogue multiplication. I think the relevant code is here? https://github.com/pytorch/pytorch/blob/90e12cf63d7ead464e9f99007b50b494e82298ca/torch/_inductor/kernel/mm.py#L757
- This might not be an issue, as the other multiplication can be fused into the subsequent op. However, in a small end2end benchmark, my explicit triton kernel is still much faster (50%).

Therefore, I think for now using the explicit triton kernel is still better.

What I'm curious is that is there a mechanism to register a custom op with torch compiler that provides 2 implementations (e.g. 1 hand-written fused Triton impl, and 1 pure PyTorch impl which can potentially be fused across op boundaries) so that the compiler can benchmark and choose the better one in a specific scenario.